Memory pair game by Evgen24

Evgeniy241984 opened the pull request:
### Memory-Pair-Game

[Demo](https://evgeniy241984.github.io/memory-pair-game/)
[Code base](https://github.com/Evgeniy241984/frontend-2021-homeworks/blob/memory-pair-game/submissions/evgeniy24/memory-pair-game/app.js)

The code is submitted in a dedicated feature branch.
Please, review.

chris-voitova commented:
Hi! Interesting design idea! I saw that the array was never shuffled when I reload the page, this is not interesting for players. And the matched cards were not removed, but should. I also won when there was one card left. Please fix
<img width="826" alt="Screenshot 2020-11-27 at 17 37 28" src="https://user-images.githubusercontent.com/25912048/100465816-9ec96500-30d8-11eb-80db-246ba738383f.png">
<img width="821" alt="Screenshot 2020-11-27 at 17 43 11" src="https://user-images.githubusercontent.com/25912048/100465827-a2f58280-30d8-11eb-98db-7fe28adfba83.png">

Evgeniy241984 commented:
Hi, Thank you for comment! 
i fixed double click issue, added shuffling card position, and mathed cards dissappears now after several sec! 
See demo!

MaxKalashnyk commented on submissions/evgeniy24/memory-pair-game/styles.css:
```diff
@@ -0,0 +1,79 @@
+  perspective: 1500px;
+}
+
+.card {
```
Add `cusrsor: pointer;` to the cards on hover. It would improve UX of your app

Evgeniy241984 replied:
fixed! Pleasee see the result!

MaxKalashnyk replied:
I see it in your code, but don't see in your demo :(

Evgeniy241984 replied:
Fixed in Demo !

MaxKalashnyk commented on submissions/evgeniy24/memory-pair-game/app.js:
```diff
@@ -0,0 +1,81 @@
+    })
+})();
+
+cards.forEach(function(card) {
```
Use [event delegation](https://learn.javascript.ru/event-delegation), please

Evgeniy241984 replied:
fixed! Pleasee see the result!

MaxKalashnyk commented on submissions/evgeniy24/memory-pair-game/app.js:
```diff
@@ -0,0 +1,81 @@
+    };
+}
+
+(function resetOrderCards() {
```
Why did you use [IIFE](https://developer.mozilla.org/en-US/docs/Glossary/IIFE) here?

Evgeniy241984 replied:
fixed! Pleasee see the result!

MaxKalashnyk commented on submissions/evgeniy24/memory-pair-game/app.js:
```diff
@@ -0,0 +1,81 @@
+
+(function resetOrderCards() {
+    cards.forEach(function(card) {
+        let position = Math.floor(Math.random() * 12);
```
you can use const for this variable. You don't reassign it so you can use const with the confidence here

Evgeniy241984 replied:
fixed! Pleasee see the result!

MaxKalashnyk commented on submissions/evgeniy24/memory-pair-game/app.js:
```diff
@@ -0,0 +1,81 @@
+}
+
+function checkWinCondition() {
+    if (winCondition === 6) {
```
Use constant for 6, please

Evgeniy241984 replied:
fixed! Pleasee see the result!

MaxKalashnyk reviewed (changes requested):
Nice work! Look at my comments below, please

MaxKalashnyk commented on submissions/evgeniy24/memory-pair-game/app.js:
```diff
@@ -0,0 +1,89 @@
+
+window.onload = resetOrderCards();
+
+field.onclick = function(event) {
```
Use `addEventListener`, please

Evgeniy241984 replied:
Fixed, Now i using an addEventListener!

MaxKalashnyk commented on submissions/evgeniy24/memory-pair-game/app.js:
```diff
@@ -0,0 +1,89 @@
+window.onload = resetOrderCards();
+
+field.onclick = function(event) {
+    if (event.target.className != 'back') return;
```
Use `classList.contains`, please

Evgeniy241984 replied:
Done! Using classList.contains!

MaxKalashnyk commented on submissions/evgeniy24/memory-pair-game/app.js:
```diff
@@ -0,0 +1,89 @@
+field.onclick = function(event) {
+    if (event.target.className != 'back') return;
+    
+    let card = event.target.closest('.card');
```
const, please

Evgeniy241984 replied:
OK, fixed!

MaxKalashnyk commented on submissions/evgeniy24/memory-pair-game/app.js:
```diff
@@ -0,0 +1,89 @@
+
+function resetOrderCards() {
+    cards.forEach(function(card) {
+        let position = Math.floor(Math.random() * cardMaxPositions);
```
use const, please. You didn't do any of the reassignments here

Evgeniy241984 replied:
Done! using const!

MaxKalashnyk commented on submissions/evgeniy24/memory-pair-game/app.js:
```diff
@@ -0,0 +1,89 @@
+let isRolledCard = false,
+    lockRoll = false,
+    firstCard, secondCard,
+    winCondition = 6,
```
It's better to make them constants in order to prevent reassignment in your further code

Evgeniy241984 replied:
I use const for winConditions and cardMaxPositions because they are not reassigned anywhere further! I use let for other variables because they are!

MaxKalashnyk reviewed (approved):
LGTM! Good job!
